# Alignements Athena BDPM - RxNorm - Quelques stats

In [1]:
import pandas as pd

In [2]:
# Table concept
concept = pd.read_csv('concept.csv', sep='\t')
concept.head()

C:\Users\GaëlleRimaud\AppData\Local\Temp\ipykernel_17428\3722695219.py:1: DtypeWarning: Columns (5,6,9) have mixed types. Specify dtype option on import or set low_memory=False.
  concept = pd.read_csv('concept.csv', sep='\t')


concept_id                             concept_name domain_id  \
0    45756805                     Pediatric Cardiology  Provider   
1    45756804                 Pediatric Anesthesiology  Provider   
2    45756803  Pathology-Anatomic / Pathology-Clinical  Provider   
3    45756802                    Pathology - Pediatric  Provider   
4    45756801            Pathology - Molecular Genetic  Provider   

  vocabulary_id     concept_class_id standard_concept concept_code  \
0          ABMS  Physician Specialty                S  OMOP4821938   
1          ABMS  Physician Specialty                S  OMOP4821939   
2          ABMS  Physician Specialty                S  OMOP4821940   
3          ABMS  Physician Specialty                S  OMOP4821941   
4          ABMS  Physician Specialty                S  OMOP4821942   

   valid_start_date  valid_end_date invalid_reason  
0          19700101        20991231            NaN  
1          19700101        20991231            NaN  
2          19700101        20991231            NaN  
3          19700101        20991231            NaN  
4          19700101        20991231            NaN

In [11]:
# Table concept_relationship --> Table qui contient les relations entre les codes (alignements, hiérarchiques, etc)
concept_rel = pd.read_csv('CONCEPT_RELATIONSHIP.csv', sep='\t')
concept_rel.head()

concept_id_1  concept_id_2      relationship_id  valid_start_date  \
0      42628856      42628856          Mapped from          20160606   
1      42628856      42628856              Maps to          20160606   
2      42628859      40165064  Has quantified form          20160606   
3      42628859      42628859          Mapped from          20160606   
4      42628859      42628859              Maps to          20160606   

   valid_end_date  invalid_reason  
0        20991231             NaN  
1        20991231             NaN  
2        20991231             NaN  
3        20991231             NaN  
4        20991231             NaN

In [3]:
# Nomenclatures uniques dans concept
concept.vocabulary_id.unique()

array(['ABMS', 'SNOMED', 'ATC', 'BDPM', 'CCAM', 'CDM', 'CIM10',
       'CMS Place of Service', 'Concept Class', 'Condition Status',
       'Condition Type', 'Cost', 'Cost Type', 'Death Type', 'Device Type',
       'Domain', 'Drug Type', 'Episode', 'RxNorm', 'Gender', 'ICD10CM',
       'ICD10PCS', 'Korean Revenue Code', 'LOINC', 'Language',
       'Relationship', 'Meas Type', 'Medicare Specialty', 'Metadata',
       'None', 'Note Type', 'OSM', 'Obs Period Type', 'Observation Type',
       'Plan', 'Plan Stop Reason', 'Procedure Type', 'SOPT', 'Sponsor',
       'Type Concept', 'UB04 Point of Origin', 'UB04 Pri Typ of Adm',
       'UB04 Pt dis status', 'UB04 Typ bill', 'UCUM', 'US Census',
       'Visit', 'Visit Type', 'Vocabulary'], dtype=object)

### Table concept pour les codes de la BDPM

In [50]:
# Table concept pour les codes de la BDPM
bdpm = concept[concept.vocabulary_id=='BDPM'][['concept_code','concept_id','concept_name','concept_class_id']].rename(columns={'concept_code' : 'bdpm_code', 'concept_id': 'bdpm_concept_id','concept_name':'bdpm_libelle','concept_class_id':'bdpm_concept_class_id'})
bdpm.head()

bdpm_code  bdpm_concept_id           bdpm_libelle bdpm_concept_class_id
6915     00001         43047359             Loperamide            Ingredient
6916     00005         43045429   Acetylsalicylic Acid            Ingredient
6917     00007         43045800           Benzoic Acid            Ingredient
6918     00008         43045866             Boric Acid            Ingredient
6919     00011         43045611  Anhydrous Citric Acid            Ingredient

In [51]:
#Nombre de valeurs uniques
bdpm.nunique()

bdpm_code                44376
bdpm_concept_id          44376
bdpm_libelle             31140
bdpm_concept_class_id        6
dtype: int64

In [53]:
# Nombre de lignes par classe
bdpm.bdpm_concept_class_id.value_counts()

Drug Product    22113
Brand Name      15053
Ingredient       4310
Supplier         1504
Device           1162
Dose Form         234
Name: bdpm_concept_class_id, dtype: int64

### Table concept pour les codes RxNorm

In [54]:
rxnorm = concept[concept.vocabulary_id=='RxNorm'][['concept_code','concept_id','concept_name','concept_class_id']].rename(columns={'concept_code':'rxnorm_code','concept_name':'rxnorm_libelle','concept_id' : 'rxnorm_concept_id','concept_class_id':'rxnorm_concept_class_id'})
rxnorm.head()

rxnorm_code  rxnorm_concept_id  \
63745     1042838           40229458   
63746     1042847           40229459   
63747     1049224           40232055   
63748     1371295           43525657   
63749     1371298           43525658   

                                          rxnorm_libelle  \
63745                     benzocaine 0.18 MG/MG Oral Gel   
63746                benzocaine 0.18 MG/MG Oral Ointment   
63747     benzocaine 0.18 MG/MG Oral Ointment [Topicale]   
63748  benzocaine 180 MG/ML Mucous Membrane Topical S...   
63749  benzocaine 180 MG/ML Mucous Membrane Topical S...   

      rxnorm_concept_class_id  
63745           Clinical Drug  
63746           Clinical Drug  
63747            Branded Drug  
63748           Clinical Drug  
63749            Branded Drug

In [55]:
# Nombre de valeurs uniques
rxnorm.nunique()

rxnorm_code                308709
rxnorm_concept_id          308709
rxnorm_libelle             306364
rxnorm_concept_class_id        18
dtype: int64

In [61]:
# Nombre de codes par classe
rxnorm.rxnorm_concept_class_id.value_counts()

Clinical Drug           53083
Branded Drug            39083
Clinical Drug Comp      38474
Branded Drug Comp       36748
Branded Drug Form       26595
Branded Dose Group      23388
Clinical Drug Form      20773
Brand Name              19469
Clinical Dose Group     17741
Ingredient              15604
Multiple Ingredients     4469
Quant Clinical Drug      3860
Precise Ingredient       3476
Quant Branded Drug       3215
Branded Pack             1308
Clinical Pack            1175
Dose Form                 201
Dose Form Group            47
Name: rxnorm_concept_class_id, dtype: int64

### Codes de la bdpm dans la table des relations

In [64]:
bdpm_concept_rel = bdpm.merge(concept_rel, left_on=['bdpm_concept_id'], right_on=['concept_id_1'])
bdpm_concept_rel.head()

bdpm_code  bdpm_concept_id           bdpm_libelle bdpm_concept_class_id  \
0     00001         43047359             Loperamide            Ingredient   
1     00005         43045429   Acetylsalicylic Acid            Ingredient   
2     00007         43045800           Benzoic Acid            Ingredient   
3     00008         43045866             Boric Acid            Ingredient   
4     00011         43045611  Anhydrous Citric Acid            Ingredient   

   concept_id_1  concept_id_2 relationship_id  valid_start_date  \
0      43047359        991876         Maps to          20191006   
1      43045429       1112807         Maps to          20191006   
2      43045800        711452         Maps to          20191006   
3      43045866        929698         Maps to          20191006   
4      43045611        950435         Maps to          20191006   

   valid_end_date  invalid_reason  
0        20991231             NaN  
1        20991231             NaN  
2        20991231             NaN  
3        20991231             NaN  
4        20991231             NaN

In [65]:
# Valuers uniques par colonne
bdpm_concept_rel.nunique()

bdpm_code                7300
bdpm_concept_id          7300
bdpm_libelle             5949
bdpm_concept_class_id       5
concept_id_1             7300
concept_id_2             4445
relationship_id             3
valid_start_date            4
valid_end_date              1
invalid_reason              0
dtype: int64

Seulement 7300 / 44376 codes de la bdpm sont dans concept_relationship.

### Relations entre les codes de la BDPM et RxNorm

In [67]:
bdpm_rxnorm = bdpm_concept_rel.merge(rxnorm, left_on=['concept_id_2'], right_on=['rxnorm_concept_id']).drop(columns=['valid_start_date','valid_end_date','invalid_reason','concept_id_1','concept_id_2'])
bdpm_rxnorm.head()

bdpm_code  bdpm_concept_id                  bdpm_libelle  \
0     00001         43047359                    Loperamide   
1         1         43239732      Loperamide hydrochloride   
2         1         43239732      Loperamide hydrochloride   
3     27324         43236076  Loperamide oxide monohydrate   
4     27324         43236076  Loperamide oxide monohydrate   

  bdpm_concept_class_id     relationship_id rxnorm_code  rxnorm_concept_id  \
0            Ingredient             Maps to        6468             991876   
1            Ingredient             Maps to        6468             991876   
2            Ingredient  Source - RxNorm eq        6468             991876   
3            Ingredient             Maps to        6468             991876   
4            Ingredient  Source - RxNorm eq        6468             991876   

  rxnorm_libelle rxnorm_concept_class_id  
0     loperamide              Ingredient  
1     loperamide              Ingredient  
2     loperamide              Ingredient  
3     loperamide              Ingredient  
4     loperamide              Ingredient

In [68]:
# Nombre de valeurs uniques par colonnes
bdpm_rxnorm.nunique()

bdpm_code                  6194
bdpm_concept_id            6194
bdpm_libelle               5491
bdpm_concept_class_id         4
relationship_id               2
rxnorm_code                3339
rxnorm_concept_id          3339
rxnorm_libelle             3338
rxnorm_concept_class_id       9
dtype: int64

Seulement 6194 / 44 376 codes de la BDPM sont alignés avec un code RxNorm

In [59]:
# Nombre de valeurs uniques par classe de la BDPM
bdpm_rxnorm.bdpm_concept_class_id.value_counts()

Ingredient      5903
Brand Name      1805
Dose Form        231
Drug Product     184
Name: bdpm_concept_class_id, dtype: int64

In [60]:
# Nombre de valeurs unique par classe RxNorm
bdpm_rxnorm.rxnorm_concept_class_id.value_counts()

Ingredient             5879
Brand Name             1805
Dose Form               231
Clinical Drug           139
Precise Ingredient       24
Clinical Drug Form       22
Clinical Pack            15
Quant Clinical Drug       6
Branded Drug Form         2
Name: rxnorm_concept_class_id, dtype: int64

In [66]:
# Nombre de valeurs uniques par type de relation
bdpm_rxnorm.relationship_id.value_counts()

Maps to               4101
Source - RxNorm eq    4022
Name: relationship_id, dtype: int64

In [42]:
# Nombre de relations 'Maps to'
bdpm_rxnorm[bdpm_rxnorm.relationship_id=='Maps to'].nunique()

bdpm_code            4101
bdpm_concept_id      4101
bdpm_libelle         3540
relationship_id         1
rxnorm_code          2232
rxnorm_concept_id    2232
rxnorm_libelle       2232
dtype: int64

In [70]:
bdpm_rxnorm.to_csv('mapping_athena_bdpm_rxnorm.csv', index=False)